File: day_25_analytical_windows/day_25_analytical_windows.js
/*
 * Analytical Windows: LAG, LEAD, FIRST_VALUE, LAST_VALUE
 * ========================================================
 *
 * A self-contained JavaScript study file.
 *
 * JavaScript does not provide SQL window functions as core language
 * constructs. This file therefore implements the underlying analytical
 * mechanics with arrays, sorting, partitioning, and reusable functions.
 *
 * The implementations model the semantics of:
 *   LAG()
 *   LEAD()
 *   FIRST_VALUE()
 *   LAST_VALUE()
 *
 * The file also demonstrates:
 *   - partitions
 *   - deterministic ordering
 *   - offsets
 *   - defaults
 *   - NULL-like missing values
 *   - frame concepts
 *   - customer journey analysis
 *   - percentage changes
 *   - activity gaps
 *   - validation
 *   - performance considerations
 */

"use strict";

// -----------------------------------------------------------------------------
// 1. SAMPLE DATA
// -----------------------------------------------------------------------------

const sales = [
    { saleId: 1, customerId: 101, region: "North", date: "2026-01-05", product: "Laptop", quantity: 1, unitPrice: 1200 },
    { saleId: 2, customerId: 101, region: "North", date: "2026-01-20", product: "Mouse", quantity: 2, unitPrice: 25 },
    { saleId: 3, customerId: 101, region: "North", date: "2026-02-03", product: "Monitor", quantity: 1, unitPrice: 350 },
    { saleId: 4, customerId: 101, region: "North", date: "2026-02-20", product: "Keyboard", quantity: 1, unitPrice: 80 },
    { saleId: 5, customerId: 101, region: "North", date: "2026-03-10", product: "Laptop", quantity: 1, unitPrice: 1250 },

    { saleId: 6, customerId: 102, region: "South", date: "2026-01-07", product: "Phone", quantity: 1, unitPrice: 800 },
    { saleId: 7, customerId: 102, region: "South", date: "2026-01-25", product: "Case", quantity: 2, unitPrice: 30 },
    { saleId: 8, customerId: 102, region: "South", date: "2026-02-11", product: "Phone", quantity: 1, unitPrice: 820 },
    { saleId: 9, customerId: 102, region: "South", date: "2026-03-05", product: "Earbuds", quantity: 1, unitPrice: 150 },

    { saleId: 10, customerId: 103, region: "North", date: "2026-01-15", product: "Tablet", quantity: 1, unitPrice: 500 },
    { saleId: 11, customerId: 103, region: "North", date: "2026-02-15", product: "Tablet", quantity: 1, unitPrice: 500 },
    { saleId: 12, customerId: 103, region: "North", date: "2026-03-15", product: "Stylus", quantity: 1, unitPrice: 60 },

    { saleId: 13, customerId: 104, region: "West", date: "2026-01-03", product: "Camera", quantity: 1, unitPrice: 1000 },
    { saleId: 14, customerId: 104, region: "West", date: "2026-01-03", product: "Tripod", quantity: 1, unitPrice: 120 },
    { saleId: 15, customerId: 104, region: "West", date: "2026-02-01", product: "Lens", quantity: 1, unitPrice: 700 },
    { saleId: 16, customerId: 104, region: "West", date: "2026-03-01", product: "Camera", quantity: 1, unitPrice: 1100 },

    { saleId: 17, customerId: 105, region: "East", date: "2026-02-01", product: "Chair", quantity: 1, unitPrice: 250 },
    { saleId: 18, customerId: 105, region: "East", date: "2026-03-01", product: "Desk", quantity: 1, unitPrice: 450 }
];

function withRevenue(row) {
    return {
        ...row,
        revenue: row.quantity * row.unitPrice
    };
}

const salesWithRevenue = sales.map(withRevenue);

// -----------------------------------------------------------------------------
// 2. GENERAL UTILITIES
// -----------------------------------------------------------------------------

function cloneRows(rows) {
    return rows.map(row => ({ ...row }));
}

function compareByDateThenId(a, b) {
    const dateComparison = a.date.localeCompare(b.date);

    if (dateComparison !== 0) {
        return dateComparison;
    }

    return a.saleId - b.saleId;
}

function partitionBy(rows, keySelector) {
    const partitions = new Map();

    for (const row of rows) {
        const key = keySelector(row);

        if (!partitions.has(key)) {
            partitions.set(key, []);
        }

        partitions.get(key).push(row);
    }

    return partitions;
}

function orderedPartitions(rows, partitionSelector, comparator) {
    const partitions = partitionBy(rows, partitionSelector);

    for (const partition of partitions.values()) {
        partition.sort(comparator);
    }

    return partitions;
}

function printRows(rows, columns) {
    console.table(
        rows.map(row => {
            const result = {};

            for (const column of columns) {
                result[column] = row[column];
            }

            return result;
        })
    );
}

// -----------------------------------------------------------------------------
// 3. CORE WINDOW OPERATIONS
// -----------------------------------------------------------------------------

function lag(values, offset = 1, defaultValue = null) {
    if (!Number.isInteger(offset) || offset < 0) {
        throw new RangeError("LAG offset must be a non-negative integer.");
    }

    return values.map((_, index) => {
        const sourceIndex = index - offset;

        return sourceIndex >= 0
            ? values[sourceIndex]
            : defaultValue;
    });
}

function lead(values, offset = 1, defaultValue = null) {
    if (!Number.isInteger(offset) || offset < 0) {
        throw new RangeError("LEAD offset must be a non-negative integer.");
    }

    return values.map((_, index) => {
        const sourceIndex = index + offset;

        return sourceIndex < values.length
            ? values[sourceIndex]
            : defaultValue;
    });
}

function firstValue(values, defaultValue = null) {
    return values.length > 0 ? values[0] : defaultValue;
}

function lastValue(values, defaultValue = null) {
    return values.length > 0
        ? values[values.length - 1]
        : defaultValue;
}

// -----------------------------------------------------------------------------
// 4. SQL-LIKE PARTITIONED WINDOW FUNCTIONS
// -----------------------------------------------------------------------------

function applyPartitionedWindow(
    rows,
    partitionSelector,
    orderComparator,
    valueSelector
) {
    const partitions = orderedPartitions(
        rows,
        partitionSelector,
        orderComparator
    );

    const result = [];

    for (const partition of partitions.values()) {
        const values = partition.map(valueSelector);

        const previousValues = lag(values);
        const nextValues = lead(values);
        const first = firstValue(values);
        const last = lastValue(values);

        partition.forEach((row, index) => {
            result.push({
                ...row,
                previousValue: previousValues[index],
                nextValue: nextValues[index],
                firstValue: first,
                lastValue: last
            });
        });
    }

    result.sort((a, b) => {
        const customerComparison = a.customerId - b.customerId;

        if (customerComparison !== 0) {
            return customerComparison;
        }

        return compareByDateThenId(a, b);
    });

    return result;
}

// -----------------------------------------------------------------------------
// 5. LAG DEMONSTRATION
// -----------------------------------------------------------------------------

function demonstrateLag() {
    console.log("\n1. LAG(): previous sale revenue");

    const result = applyPartitionedWindow(
        salesWithRevenue,
        row => row.customerId,
        compareByDateThenId,
        row => row.revenue
    );

    printRows(result, [
        "saleId",
        "customerId",
        "date",
        "revenue",
        "previousValue"
    ]);
}

// -----------------------------------------------------------------------------
// 6. LEAD DEMONSTRATION
// -----------------------------------------------------------------------------

function demonstrateLead() {
    console.log("\n2. LEAD(): next sale revenue");

    const result = applyPartitionedWindow(
        salesWithRevenue,
        row => row.customerId,
        compareByDateThenId,
        row => row.revenue
    );

    printRows(result, [
        "saleId",
        "customerId",
        "date",
        "revenue",
        "nextValue"
    ]);
}

// -----------------------------------------------------------------------------
// 7. FIRST_VALUE AND LAST_VALUE
// -----------------------------------------------------------------------------

function demonstrateFirstAndLast() {
    console.log("\n3. FIRST_VALUE() and LAST_VALUE() semantics");

    const result = applyPartitionedWindow(
        salesWithRevenue,
        row => row.customerId,
        compareByDateThenId,
        row => row.revenue
    );

    printRows(result, [
        "saleId",
        "customerId",
        "date",
        "revenue",
        "firstValue",
        "lastValue"
    ]);
}

// -----------------------------------------------------------------------------
// 8. CUSTOMER JOURNEY
// -----------------------------------------------------------------------------

function customerJourney() {
    console.log("\n4. Customer journey analysis");

    const partitions = orderedPartitions(
        salesWithRevenue,
        row => row.customerId,
        compareByDateThenId
    );

    const result = [];

    for (const [customerId, rows] of partitions) {
        const revenues = rows.map(row => row.revenue);
        const products = rows.map(row => row.product);

        const previousRevenue = lag(revenues);
        const nextRevenue = lead(revenues);
        const previousProduct = lag(products);
        const nextProduct = lead(products);

        rows.forEach((row, index) => {
            result.push({
                customerId,
                saleDate: row.date,
                product: row.product,
                previousProduct: previousProduct[index],
                nextProduct: nextProduct[index],
                revenue: row.revenue,
                changeFromPrevious:
                    previousRevenue[index] == null
                        ? null
                        : row.revenue - previousRevenue[index],
                changeToNext:
                    nextRevenue[index] == null
                        ? null
                        : nextRevenue[index] - row.revenue
            });
        }
    }

    printRows(result, [
        "customerId",
        "saleDate",
        "product",
        "previousProduct",
        "nextProduct",
        "revenue",
        "changeFromPrevious",
        "changeToNext"
    ]);
}

// -----------------------------------------------------------------------------
// 9. PERCENTAGE CHANGE
// -----------------------------------------------------------------------------

function percentageChange(currentValue, previousValue) {
    if (previousValue == null || previousValue === 0) {
        return null;
    }

    return ((currentValue - previousValue) / previousValue) * 100;
}

function demonstratePercentageChange() {
    console.log("\n5. Percentage change");

    const partitions = orderedPartitions(
        salesWithRevenue,
        row => row.customerId,
        compareByDateThenId
    );

    const result = [];

    for (const [customerId, rows] of partitions) {
        const revenues = rows.map(row => row.revenue);
        const previousRevenues = lag(revenues);

        rows.forEach((row, index) => {
            result.push({
                customerId,
                date: row.date,
                revenue: row.revenue,
                previousRevenue: previousRevenues[index],
                percentageChange: percentageChange(
                    row.revenue,
                    previousRevenues[index]
                )
            });
        });
    }

    printRows(result, [
        "customerId",
        "date",
        "revenue",
        "previousRevenue",
        "percentageChange"
    ]);
}

// -----------------------------------------------------------------------------
// 10. ACTIVITY GAP
// -----------------------------------------------------------------------------

function daysBetween(firstDate, secondDate) {
    const first = new Date(`${firstDate}T00:00:00Z`);
    const second = new Date(`${secondDate}T00:00:00Z`);

    return Math.round(
        (second.getTime() - first.getTime()) / (24 * 60 * 60 * 1000)
    );
}

function demonstrateActivityGap() {
    console.log("\n6. Days between customer transactions");

    const partitions = orderedPartitions(
        salesWithRevenue,
        row => row.customerId,
        compareByDateThenId
    );

    const result = [];

    for (const [customerId, rows] of partitions) {
        const dates = rows.map(row => row.date);
        const previousDates = lag(dates);

        rows.forEach((row, index) => {
            result.push({
                customerId,
                date: row.date,
                previousDate: previousDates[index],
                daysSincePrevious:
                    previousDates[index] == null
                        ? null
                        : daysBetween(previousDates[index], row.date)
            });
        }
    }

    printRows(result, [
        "customerId",
        "date",
        "previousDate",
        "daysSincePrevious"
    ]);
}

// -----------------------------------------------------------------------------
// 11. OFFSET AND DEFAULT VALUES
// -----------------------------------------------------------------------------

function demonstrateOffsets() {
    console.log("\n7. LAG and LEAD with offsets and defaults");

    const values = [100, 200, 300, 400];

    console.log("Values:", values);
    console.log("LAG(values, 2, 0):", lag(values, 2, 0));
    console.log("LEAD(values, 2, 0):", lead(values, 2, 0));
}

// -----------------------------------------------------------------------------
// 12. WINDOW FRAME MODEL
// -----------------------------------------------------------------------------

function boundedAverage(values, preceding, following) {
    if (!Number.isInteger(preceding) || preceding < 0) {
        throw new RangeError("preceding must be a non-negative integer.");
    }

    if (!Number.isInteger(following) || following < 0) {
        throw new RangeError("following must be a non-negative integer.");
    }

    return values.map((_, index) => {
        const start = Math.max(0, index - preceding);
        const end = Math.min(values.length - 1, index + following);

        const frame = values.slice(start, end + 1);
        const total = frame.reduce((sum, value) => sum + value, 0);

        return total / frame.length;
    });
}

function demonstrateFrame() {
    console.log("\n8. Bounded frame model");

    const values = [100, 200, 300, 400, 500];

    console.log(
        "Centered one-row frame:",
        boundedAverage(values, 1, 1)
    );
}

// -----------------------------------------------------------------------------
// 13. DETERMINISTIC ORDERING
// -----------------------------------------------------------------------------

function demonstrateTieBreaking() {
    console.log("\n9. Deterministic ordering");

    const customerRows = salesWithRevenue
        .filter(row => row.customerId === 104)
        .sort(compareByDateThenId);

    const products = customerRows.map(row => row.product);

    printRows(
        customerRows.map((row, index) => ({
            saleId: row.saleId,
            date: row.date,
            product: row.product,
            previousProduct: lag(products)[index]
        })),
        [
            "saleId",
            "date",
            "product",
            "previousProduct"
        ]
    );
}

// -----------------------------------------------------------------------------
// 14. CLASSIFY MOVEMENT
// -----------------------------------------------------------------------------

function classifyMovement(current, previous) {
    if (previous == null) {
        return "FIRST SALE";
    }

    if (current > previous) {
        return "INCREASE";
    }

    if (current < previous) {
        return "DECREASE";
    }

    return "UNCHANGED";
}

function demonstrateMovementClassification() {
    console.log("\n10. Revenue movement classification");

    const partitions = orderedPartitions(
        salesWithRevenue,
        row => row.customerId,
        compareByDateThenId
    );

    const result = [];

    for (const [customerId, rows] of partitions) {
        const revenues = rows.map(row => row.revenue);
        const previousRevenues = lag(revenues);

        rows.forEach((row, index) => {
            result.push({
                customerId,
                date: row.date,
                revenue: row.revenue,
                previousRevenue: previousRevenues[index],
                movement: classifyMovement(
                    row.revenue,
                    previousRevenues[index]
                )
            });
        });
    }

    printRows(result, [
        "customerId",
        "date",
        "revenue",
        "previousRevenue",
        "movement"
    ]);
}

// -----------------------------------------------------------------------------
// 15. VALIDATION
// -----------------------------------------------------------------------------

function assertEqual(actual, expected, message) {
    if (actual !== expected) {
        throw new Error(
            `${message}: expected ${expected}, received ${actual}`
        );
    }
}

function assertArrayEqual(actual, expected, message) {
    if (actual.length !== expected.length) {
        throw new Error(
            `${message}: different lengths`
        );
    }

    actual.forEach((value, index) => {
        if (value !== expected[index]) {
            throw new Error(
                `${message}: index ${index}, expected ${expected[index]}, received ${value}`
            );
        }
    });
}

function runTests() {
    console.log("\n11. Automated validation");

    assertArrayEqual(
        lag([10, 20, 30]),
        [null, 10, 20],
        "LAG basic behavior"
    );

    assertArrayEqual(
        lead([10, 20, 30]),
        [20, 30, null],
        "LEAD basic behavior"
    );

    assertArrayEqual(
        lag([10, 20, 30], 2, 0),
        [0, 0, 10],
        "LAG offset/default behavior"
    );

    assertArrayEqual(
        lead([10, 20, 30], 2, 0),
        [30, 0, 0],
        "LEAD offset/default behavior"
    );

    assertEqual(
        firstValue([10, 20, 30]),
        10,
        "FIRST_VALUE behavior"
    );

    assertEqual(
        lastValue([10, 20, 30]),
        30,
        "LAST_VALUE behavior"
    );

    assertEqual(
        percentageChange(120, 100),
        20,
        "Percentage change"
    );

    assertEqual(
        percentageChange(100, 0),
        null,
        "Zero-denominator protection"
    );

    console.log("All JavaScript assertions passed.");
}

// -----------------------------------------------------------------------------
// 16. PERFORMANCE NOTES AS EXECUTABLE INFORMATION
// -----------------------------------------------------------------------------

function printPerformanceNotes() {
    console.log(`
Performance considerations
---------------------------
1. Partitioning groups rows before relative-row calculations.
2. Sorting each partition is normally the expensive operation.
3. A typical ordered analytical workflow is dominated by sorting and memory.
4. SQL databases can use indexes to reduce sorting work.
5. In JavaScript, avoid repeatedly sorting the same partition.
6. Reuse partitioned and ordered data when multiple window calculations
   use the same partition/order definition.
7. LAG and LEAD themselves are O(1) per row after ordering.
8. FIRST_VALUE and LAST_VALUE are O(1) per row after a partition is ordered.
9. Materializing large datasets increases memory usage.
10. Streaming systems require specialized state management because a
    future row is not necessarily available when the current row arrives.
`);
}

// -----------------------------------------------------------------------------
// 17. MAIN
// -----------------------------------------------------------------------------

function main() {
    console.log("Analytical Windows: LAG, LEAD, FIRST_VALUE, LAST_VALUE");

    demonstrateLag();
    demonstrateLead();
    demonstrateFirstAndLast();
    customerJourney();
    demonstratePercentageChange();
    demonstrateActivityGap();
    demonstrateOffsets();
    demonstrateFrame();
    demonstrateTieBreaking();
    demonstrateMovementClassification();
    runTests();
    printPerformanceNotes();

    console.log("\nStudy file completed successfully.");
}

main();

File: day_25_analytical_windows/day_25_analytical_windows.cpp
/*
 * Analytical Windows: LAG, LEAD, FIRST_VALUE, LAST_VALUE
 * ========================================================
 *
 * C++17 case study:
 *
 * A transaction analytics engine for a retail organization.
 *
 * The program models the analytical semantics commonly implemented in SQL
 * window functions. It progressively develops:
 *
 *   1. Transaction storage
 *   2. Partitioning by customer
 *   3. Deterministic ordering
 *   4. LAG
 *   5. LEAD
 *   6. FIRST_VALUE
 *   7. LAST_VALUE
 *   8. Percentage-change analysis
 *   9. Activity-gap analysis
 *  10. Customer lifecycle analysis
 *  11. Validation
 *  12. Complexity and architectural considerations
 *
 * Build:
 *   g++ -std=c++17 -O2 -Wall -Wextra -pedantic analytical_windows.cpp -o analytical_windows
 *
 * Run:
 *   ./analytical_windows
 */

#include <algorithm>
#include <cassert>
#include <cmath>
#include <iomanip>
#include <iostream>
#include <map>
#include <optional>
#include <stdexcept>
#include <string>
#include <tuple>
#include <unordered_map>
#include <utility>
#include <vector>

using namespace std;

// -----------------------------------------------------------------------------
// 1. DOMAIN MODEL
// -----------------------------------------------------------------------------

struct Sale {
    int saleId;
    int customerId;
    string region;
    string date;
    string product;
    int quantity;
    double unitPrice;

    double revenue() const {
        return quantity * unitPrice;
    }
};

struct CustomerAnalysis {
    int customerId;
    string date;
    string product;
    double revenue;

    optional<double> previousRevenue;
    optional<double> nextRevenue;

    optional<string> previousProduct;
    optional<string> nextProduct;

    optional<double> firstRevenue;
    optional<double> lastRevenue;

    optional<double> percentageChange;
    optional<int> daysSincePrevious;

    string movement;
};

// -----------------------------------------------------------------------------
// 2. VALIDATION
// -----------------------------------------------------------------------------

void validateSale(const Sale& sale) {
    if (sale.saleId <= 0) {
        throw invalid_argument("saleId must be positive");
    }

    if (sale.customerId <= 0) {
        throw invalid_argument("customerId must be positive");
    }

    if (sale.quantity <= 0) {
        throw invalid_argument("quantity must be positive");
    }

    if (sale.unitPrice < 0) {
        throw invalid_argument("unitPrice cannot be negative");
    }

    if (sale.date.empty()) {
        throw invalid_argument("date cannot be empty");
    }

    if (sale.product.empty()) {
        throw invalid_argument("product cannot be empty");
    }
}

// -----------------------------------------------------------------------------
// 3. DATE UTILITIES
// -----------------------------------------------------------------------------

struct SimpleDate {
    int year;
    int month;
    int day;
};

SimpleDate parseDate(const string& value) {
    if (value.size() != 10 ||
        value[4] != '-' ||
        value[7] != '-') {
        throw invalid_argument("Date must use YYYY-MM-DD format: " + value);
    }

    return {
        stoi(value.substr(0, 4)),
        stoi(value.substr(5, 2)),
        stoi(value.substr(8, 2))
    };
}

bool isLeapYear(int year) {
    return (year % 400 == 0) ||
           (year % 4 == 0 && year % 100 != 0);
}

int daysInMonth(int year, int month) {
    static const int days[] = {
        31, 28, 31, 30, 31, 30,
        31, 31, 30, 31, 30, 31
    };

    if (month == 2 && isLeapYear(year)) {
        return 29;
    }

    return days[month - 1];
}

long long daysFromCivil(int year, unsigned month, unsigned day) {
    year -= month <= 2;

    const long long era =
        (year >= 0 ? year : year - 399) / 400;

    const unsigned yearOfEra =
        static_cast<unsigned>(year - era * 400);

    const unsigned dayOfYear =
        (153 * (month + (month > 2 ? -3 : 9)) + 2) / 5
        + day - 1;

    const unsigned dayOfEra =
        yearOfEra * 365
        + yearOfEra / 4
        - yearOfEra / 100
        + dayOfYear;

    return era * 146097
         + static_cast<long long>(dayOfEra)
         - 719468;
}

int daysBetween(const string& earlier, const string& later) {
    const SimpleDate a = parseDate(earlier);
    const SimpleDate b = parseDate(later);

    const long long first =
        daysFromCivil(a.year, a.month, a.day);

    const long long second =
        daysFromCivil(b.year, b.month, b.day);

    return static_cast<int>(second - first);
}

// -----------------------------------------------------------------------------
// 4. DATASET
// -----------------------------------------------------------------------------

vector<Sale> createSales() {
    vector<Sale> sales = {
        {1, 101, "North", "2026-01-05", "Laptop", 1, 1200.00},
        {2, 101, "North", "2026-01-20", "Mouse", 2, 25.00},
        {3, 101, "North", "2026-02-03", "Monitor", 1, 350.00},
        {4, 101, "North", "2026-02-20", "Keyboard", 1, 80.00},
        {5, 101, "North", "2026-03-10", "Laptop", 1, 1250.00},

        {6, 102, "South", "2026-01-07", "Phone", 1, 800.00},
        {7, 102, "South", "2026-01-25", "Case", 2, 30.00},
        {8, 102, "South", "2026-02-11", "Phone", 1, 820.00},
        {9, 102, "South", "2026-03-05", "Earbuds", 1, 150.00},

        {10, 103, "North", "2026-01-15", "Tablet", 1, 500.00},
        {11, 103, "North", "2026-02-15", "Tablet", 1, 500.00},
        {12, 103, "North", "2026-03-15", "Stylus", 1, 60.00},

        {13, 104, "West", "2026-01-03", "Camera", 1, 1000.00},
        {14, 104, "West", "2026-01-03", "Tripod", 1, 120.00},
        {15, 104, "West", "2026-02-01", "Lens", 1, 700.00},
        {16, 104, "West", "2026-03-01", "Camera", 1, 1100.00},

        {17, 105, "East", "2026-02-01", "Chair", 1, 250.00},
        {18, 105, "East", "2026-03-01", "Desk", 1, 450.00}
    };

    for (const Sale& sale : sales) {
        validateSale(sale);
    }

    return sales;
}

// -----------------------------------------------------------------------------
// 5. DETERMINISTIC ORDERING
// -----------------------------------------------------------------------------

bool saleOrder(const Sale& a, const Sale& b) {
    if (a.customerId != b.customerId) {
        return a.customerId < b.customerId;
    }

    if (a.date != b.date) {
        return a.date < b.date;
    }

    // saleId makes ordering deterministic when dates are equal.
    return a.saleId < b.saleId;
}

// -----------------------------------------------------------------------------
// 6. PARTITIONING
// -----------------------------------------------------------------------------

map<int, vector<Sale>> partitionByCustomer(
    const vector<Sale>& sales
) {
    map<int, vector<Sale>> partitions;

    for (const Sale& sale : sales) {
        partitions[sale.customerId].push_back(sale);
    }

    for (auto& [customerId, rows] : partitions) {
        sort(
            rows.begin(),
            rows.end(),
            [](const Sale& a, const Sale& b) {
                if (a.date != b.date) {
                    return a.date < b.date;
                }

                return a.saleId < b.saleId;
            }
        );
    }

    return partitions;
}

// -----------------------------------------------------------------------------
// 7. GENERIC LAG
// -----------------------------------------------------------------------------

template <typename T>
optional<T> lag(
    const vector<T>& values,
    size_t index,
    size_t offset = 1
) {
    if (offset > index) {
        return nullopt;
    }

    return values[index - offset];
}

// -----------------------------------------------------------------------------
// 8. GENERIC LEAD
// -----------------------------------------------------------------------------

template <typename T>
optional<T> lead(
    const vector<T>& values,
    size_t index,
    size_t offset = 1
) {
    if (index + offset >= values.size()) {
        return nullopt;
    }

    return values[index + offset];
}

// -----------------------------------------------------------------------------
// 9. FIRST_VALUE
// -----------------------------------------------------------------------------

template <typename T>
optional<T> firstValue(const vector<T>& values) {
    if (values.empty()) {
        return nullopt;
    }

    return values.front();
}

// -----------------------------------------------------------------------------
// 10. LAST_VALUE
// -----------------------------------------------------------------------------

template <typename T>
optional<T> lastValue(const vector<T>& values) {
    if (values.empty()) {
        return nullopt;
    }

    return values.back();
}

// -----------------------------------------------------------------------------
// 11. MOVEMENT CLASSIFICATION
// -----------------------------------------------------------------------------

string classifyMovement(
    double current,
    const optional<double>& previous
) {
    if (!previous.has_value()) {
        return "FIRST SALE";
    }

    if (current > previous.value()) {
        return "INCREASE";
    }

    if (current < previous.value()) {
        return "DECREASE";
    }

    return "UNCHANGED";
}

// -----------------------------------------------------------------------------
// 12. PERCENTAGE CHANGE
// -----------------------------------------------------------------------------

optional<double> percentageChange(
    double current,
    const optional<double>& previous
) {
    if (!previous.has_value()) {
        return nullopt;
    }

    if (previous.value() == 0.0) {
        // Division by zero is not analytically meaningful here.
        return nullopt;
    }

    return ((current - previous.value()) / previous.value()) * 100.0;
}

// -----------------------------------------------------------------------------
// 13. CUSTOMER ANALYTICS ENGINE
// -----------------------------------------------------------------------------

class CustomerAnalyticsEngine {
private:
    vector<Sale> sales;

public:
    explicit CustomerAnalyticsEngine(vector<Sale> input)
        : sales(std::move(input)) {}

    vector<CustomerAnalysis> analyze() const {
        const auto partitions = partitionByCustomer(sales);

        vector<CustomerAnalysis> result;

        for (const auto& [customerId, rows] : partitions) {
            vector<double> revenues;
            vector<string> products;
            vector<string> dates;

            revenues.reserve(rows.size());
            products.reserve(rows.size());
            dates.reserve(rows.size());

            for (const Sale& sale : rows) {
                revenues.push_back(sale.revenue());
                products.push_back(sale.product);
                dates.push_back(sale.date);
            }

            const optional<double> firstRevenue =
                firstValue(revenues);

            const optional<double> lastRevenue =
                lastValue(revenues);

            for (size_t index = 0; index < rows.size(); ++index) {
                CustomerAnalysis analysis;

                analysis.customerId = customerId;
                analysis.date = rows[index].date;
                analysis.product = rows[index].product;
                analysis.revenue = rows[index].revenue();

                analysis.previousRevenue =
                    lag(revenues, index);

                analysis.nextRevenue =
                    lead(revenues, index);

                analysis.previousProduct =
                    lag(products, index);

                analysis.nextProduct =
                    lead(products, index);

                analysis.firstRevenue = firstRevenue;
                analysis.lastRevenue = lastRevenue;

                analysis.percentageChange =
                    percentageChange(
                        analysis.revenue,
                        analysis.previousRevenue
                    );

                if (index > 0) {
                    analysis.daysSincePrevious =
                        daysBetween(
                            dates[index - 1],
                            dates[index]
                        );
                }

                analysis.movement =
                    classifyMovement(
                        analysis.revenue,
                        analysis.previousRevenue
                    );

                result.push_back(std::move(analysis));
            }
        }

        return result;
    }
};

// -----------------------------------------------------------------------------
// 14. REPORTING
// -----------------------------------------------------------------------------

string optionalNumber(const optional<double>& value) {
    if (!value.has_value()) {
        return "-";
    }

    ostringstream stream;
    stream << fixed << setprecision(2) << value.value();
    return stream.str();
}

string optionalInteger(const optional<int>& value) {
    if (!value.has_value()) {
        return "-";
    }

    return to_string(value.value());
}

string optionalText(const optional<string>& value) {
    if (!value.has_value()) {
        return "-";
    }

    return value.value();
}

void printAnalysis(
    const vector<CustomerAnalysis>& analysis
) {
    cout << "\n"
         << left
         << setw(10) << "Customer"
         << setw(13) << "Date"
         << setw(14) << "Product"
         << setw(12) << "Revenue"
         << setw(13) << "Previous"
         << setw(13) << "Next"
         << setw(13) << "First"
         << setw(13) << "Last"
         << setw(12) << "Change%"
         << setw(10) << "GapDays"
         << setw(14) << "Movement"
         << "\n";

    cout << string(147, '-') << "\n";

    for (const auto& row : analysis) {
        cout << left
             << setw(10) << row.customerId
             << setw(13) << row.date
             << setw(14) << row.product
             << setw(12) << fixed << setprecision(2) << row.revenue
             << setw(13) << optionalNumber(row.previousRevenue)
             << setw(13) << optionalNumber(row.nextRevenue)
             << setw(13) << optionalNumber(row.firstRevenue)
             << setw(13) << optionalNumber(row.lastRevenue)
             << setw(12) << optionalNumber(row.percentageChange)
             << setw(10) << optionalInteger(row.daysSincePrevious)
             << setw(14) << row.movement
             << "\n";
    }
}

// -----------------------------------------------------------------------------
// 15. CUSTOMER LIFECYCLE REPORT
// -----------------------------------------------------------------------------

void printLifecycleReport(
    const vector<CustomerAnalysis>& analysis
) {
    cout << "\nCustomer lifecycle analysis\n";
    cout << string(76, '-') << "\n";

    map<int, CustomerAnalysis> firstRows;

    for (const auto& row : analysis) {
        if (!firstRows.contains(row.customerId)) {
            firstRows[row.customerId] = row;
        }
    }

    cout << left
         << setw(12) << "Customer"
         << setw(16) << "First Revenue"
         << setw(16) << "Last Revenue"
         << setw(16) << "Difference"
         << setw(16) << "Status"
         << "\n";

    cout << string(76, '-') << "\n";

    for (const auto& [customerId, row] : firstRows) {
        const double first = row.firstRevenue.value();
        const double last = row.lastRevenue.value();
        const double difference = last - first;

        string status;

        if (difference > 0) {
            status = "Higher";
        } else if (difference < 0) {
            status = "Lower";
        } else {
            status = "Unchanged";
        }

        cout << left
             << setw(12) << customerId
             << setw(16) << fixed << setprecision(2) << first
             << setw(16) << last
             << setw(16) << difference
             << setw(16) << status
             << "\n";
    }
}

// -----------------------------------------------------------------------------
// 16. EDGE CASE TESTS
// -----------------------------------------------------------------------------

void runWindowFunctionTests() {
    cout << "\nRunning analytical-window tests...\n";

    vector<int> values = {10, 20, 30};

    assert(!lag(values, 0).has_value());
    assert(lag(values, 1).value() == 10);

    assert(lag(values, 0, 2).value() == 10);
    assert(lag(values, 1, 2).value() == 20);

    assert(!lead(values, 2).has_value());
    assert(lead(values, 0).value() == 20);
    assert(lead(values, 1).value() == 30);

    assert(firstValue(values).value() == 10);
    assert(lastValue(values).value() == 30);

    vector<int> empty;
    assert(!firstValue(empty).has_value());
    assert(!lastValue(empty).has_value());

    assert(
        percentageChange(
            120.0,
            optional<double>(100.0)
        ).value() == 20.0
    );

    assert(
        !percentageChange(
            100.0,
            optional<double>(0.0)
        ).has_value()
    );

    assert(
        daysBetween(
            "2026-01-01",
            "2026-01-31"
        ) == 30
    );

    cout << "All tests passed.\n";
}

// -----------------------------------------------------------------------------
// 17. PERFORMANCE ANALYSIS
// -----------------------------------------------------------------------------

void printPerformanceAnalysis() {
    cout << R"(
Performance model
-----------------
Let n be the number of input rows.

1. Partition construction is O(n).
2. Sorting all partitions is O(n log n) in the general case.
3. LAG and LEAD access are O(1) per row after sorting.
4. FIRST_VALUE and LAST_VALUE are O(1) per row once the partition is ordered.
5. The overall analytical pipeline is therefore generally O(n log n).
6. Memory consumption is O(n) when partitions and analytical results are
   materialized.
7. A database engine can optimize partition/order processing with indexes,
   sort operators, memory grants, parallel execution, and query planning.
8. Reusing the same partition/order definition can avoid unnecessary work.
9. A deterministic tie-breaker prevents unstable relative-row results.
10. Large analytical workloads should consider partition cardinality,
    data distribution, sorting cost, and intermediate-result size.
)";
}

// -----------------------------------------------------------------------------
// 18. DESIGN TRADE-OFFS
// -----------------------------------------------------------------------------

void printDesignTradeoffs() {
    cout << R"(
Design considerations
---------------------
LAG and LEAD:
    Best when the business question concerns relative rows.

FIRST_VALUE:
    Useful for establishing the initial state of an entity or group.

LAST_VALUE:
    Useful for final state analysis, but requires careful frame semantics
    in SQL. In this C++ model, lastValue() explicitly means the final
    element of the ordered partition.

PARTITION BY:
    Separates independent analytical histories.

ORDER BY:
    Defines the meaning of "previous", "next", "first", and "last".

Tie handling:
    Ordering only by date can be ambiguous when multiple transactions
    share the same date. saleId is therefore used as a deterministic
    secondary key.

NULL/missing values:
    The C++ optional type represents the absence of a previous or next row.
)";
}

// -----------------------------------------------------------------------------
// 19. MAIN CASE STUDY
// -----------------------------------------------------------------------------

int main() {
    try {
        cout << "Analytical Window Case Study\n";
        cout << "Retail customer transaction analytics\n";

        const vector<Sale> sales = createSales();

        CustomerAnalyticsEngine engine(sales);

        const vector<CustomerAnalysis> analysis =
            engine.analyze();

        printAnalysis(analysis);
        printLifecycleReport(analysis);
        runWindowFunctionTests();
        printPerformanceAnalysis();
        printDesignTradeoffs();

        cout << "\nCase study completed successfully.\n";
        return 0;
    }
    catch (const exception& error) {
        cerr << "Application error: "
             << error.what()
             << "\n";
        return 1;
    }
}
